app/lib/package/package_storage.dart

```diff
@@ -53,6 +56,71 @@ class PackageStorage {
     return await _publicBucket.tryInfo(objectName);
   }
 
+  /// Returns the publicly available download URL from the storage bucket.
+  Future<Uri> getPublicDownloadUrl(String package, String version) async {
+    final object = tarballObjectName(package, Uri.encodeComponent(version));
+    return Uri.parse(_publicBucket.objectUrl(object));
+  }
+
+  /// Verifies the content of an archive in the canonical bucket.
+  Future<ContentMatchStatus> matchArchiveContentInCanonical(
+    String package,
+    String version,
+    Uint8List bytes,
+  ) async {
+    final objectName = tarballObjectName(package, version);
+    final info = await _canonicalBucket.tryInfo(objectName);
+    if (info == null) {
+      return ContentMatchStatus.missing;
+    }
+    // TODO: implement quick md5 match that doesn't require to download full content
```

Comment:
This sounds important. Perhaps open an issue and attach the issue number here.

Reply:
I think this is more of a shortcut on repeated upload checks than a critical part, because it would only help the rejection path, which may be rare. Instead, I think we should move all of these checks into a separate isolate and not do it in the frontend-serving one as part of the direct request path. I'll file an issue for that and add this to it.

Reply:
I've updated the code to do the md5 hash check already with this refactor. PTAL.

Reply:
LGTM.